In [159]:
# Here we will import everything needed to do the EDA and Feature Eningeering for every dataset.
import os
import numpy as np
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
import itertools

# Import SkLearn modules for data manipluation.
from sklearn.metrics import *
from sklearn.model_selection import *
from sklearn.preprocessing import OneHotEncoder
from sklearn import preprocessing
le = preprocessing.LabelEncoder()

# Import IMUSD.
ytrain = pd.read_csv('y_train.csv')
xtrain = pd.read_csv('x_train.csv')
xtest = pd.read_csv('x_test.csv')
ytest = pd.read_csv('submission.csv')

# Import Wine Quality.
redwine = pd.read_csv('winequality-red.csv')
whitewine = pd.read_csv('winequality-white.csv')

# Import Iris
from sklearn.datasets import load_iris
irisdata = load_iris()

# Import CCD
ccdata = pd.read_csv('creditcard.csv')


FileNotFoundError: [Errno 2] File b'submission.csv' does not exist: b'submission.csv'

In [122]:
# First we will do the easy stuff.
irisdata.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [123]:
# First lets make it into a Pandas dataframe.
iris = pd.DataFrame(irisdata.data)
iris.columns = irisdata.feature_names
iristarget = pd.DataFrame(irisdata.target)
iristarget.columns = ['target']
iris.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [124]:
iristarget.groupby(['target']).count()

Empty DataFrame
Columns: []
Index: [0, 1, 2]

In [154]:
print(len(iristarget[iristarget['target']==0]))
print(len(iristarget[iristarget['target']==1]))
print(len(iristarget[iristarget['target']==2]))

50
50
50


In [126]:
# Next lets look at the data a bit.
iris.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

We see that the Iris Dataset has 150 entries with no missing values. It has 4 features with 3 equal classes. This dataset should be very easy to classify.

The next one we do will be the wine Quality datasets

In [127]:
redwine = pd.DataFrame(redwine)
redwine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [128]:
whitewine = pd.DataFrame(whitewine)
whitewine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [144]:
winedata = pd.concat([redwine, whitewine])
winedata.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [145]:
winedata['quality'] = winedata['quality'].replace(to_replace=[0,1,2,3,4,5], value=0)
winedata['quality'] = winedata['quality'].replace(to_replace=[6,7,8,9,10], value=1)
winedata.head()

In [148]:
winedata.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    6497.000000       6497.000000  6497.000000     6497.000000   
mean        7.215307          0.339666     0.318633        5.443235   
std         1.296434          0.164636     0.145318        4.757804   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.250000        1.800000   
50%         7.000000          0.290000     0.310000        3.000000   
75%         7.700000          0.400000     0.390000        8.100000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  6497.000000          6497.000000           6497.000000  6497.000000   
mean      0.056034            30.525319            115.744574     0.994697   
std       0.035034            17.749400             56.521855     0.002999   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            17.000000             77.000000     0.992340   
50%       0.047000            29.000000            118.000000     0.994890   
75%       0.065000            41.000000            156.000000     0.996990   
max       0.611000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  6497.000000  6497.000000  6497.000000  6497.000000  
mean      3.218501     0.531268    10.491801     0.633061  
std       0.160787     0.148806     1.192712     0.482007  
min       2.720000     0.220000     8.000000     0.000000  
25%       3.110000     0.430000     9.500000     0.000000  
50%       3.210000     0.510000    10.300000     1.000000  
75%       3.320000     0.600000    11.300000     1.000000  
max       4.010000     2.000000    14.900000     1.000000

We see that the wine dataset has 6497 rows with no missing values. It only has 1 classifier 'quality' which can be a 0 for bad or 1 for good. There are more bad then good but its a mostly balanced binary classifier.

Next I'm going to set up the Credit Card Data

In [152]:
ccdata.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [151]:
print(len(ccdata[ccdata["Class"]==0]))
len(ccdata[ccdata["Class"]==1])

284315


492

We see that the imbalance of the data is MASSIVE. With only 492 fraud detections we really need to do some serious manipulation to make the classifier accurate on the minority class. It will be easy for the model to just classify everything as 0 and get a 99.83% score. So in order to do this correctly we will need to get a 100% accuracy score.

Now I will finish the data munging with the IMUSD. This will require me to do a lot of feature engineering but first some EDA!

In [156]:
xtrain.head()

row_id  series_id  measurement_number  orientation_X  orientation_Y  \
0    0_0          0                   0       -0.75853       -0.63435   
1    0_1          0                   1       -0.75853       -0.63434   
2    0_2          0                   2       -0.75853       -0.63435   
3    0_3          0                   3       -0.75852       -0.63436   
4    0_4          0                   4       -0.75852       -0.63435   

   orientation_Z  orientation_W  angular_velocity_X  angular_velocity_Y  \
0       -0.10488       -0.10597            0.107650            0.017561   
1       -0.10490       -0.10600            0.067851            0.029939   
2       -0.10492       -0.10597            0.007275            0.028934   
3       -0.10495       -0.10597           -0.013053            0.019448   
4       -0.10495       -0.10596            0.005135            0.007652   

   angular_velocity_Z  linear_acceleration_X  linear_acceleration_Y  \
0            0.000767               -0.74857                 2.1030   
1            0.003385                0.33995                 1.5064   
2           -0.005978               -0.26429                 1.5922   
3           -0.008974                0.42684                 1.0993   
4            0.005245               -0.50969                 1.4689   

   linear_acceleration_Z  
0                -9.7532  
1                -9.4128  
2                -8.7267  
3               -10.0960  
4               -10.4410

In [157]:
ytrain.head()

series_id  group_id        surface
0          0        13  fine_concrete
1          1        31       concrete
2          2        20       concrete
3          3        31       concrete
4          4        22     soft_tiles

In [158]:
xtest.head()

row_id  series_id  measurement_number  orientation_X  orientation_Y  \
0    0_0          0                   0      -0.025773       -0.98864   
1    0_1          0                   1      -0.025683       -0.98862   
2    0_2          0                   2      -0.025617       -0.98861   
3    0_3          0                   3      -0.025566       -0.98862   
4    0_4          0                   4      -0.025548       -0.98866   

   orientation_Z  orientation_W  angular_velocity_X  angular_velocity_Y  \
0       -0.14801       0.003350           -0.006524           -0.001071   
1       -0.14816       0.003439           -0.113960            0.083987   
2       -0.14826       0.003571           -0.080518            0.114860   
3       -0.14817       0.003609            0.070067            0.033820   
4       -0.14792       0.003477            0.152050           -0.029016   

   angular_velocity_Z  linear_acceleration_X  linear_acceleration_Y  \
0           -0.027390                0.10043                 4.2061   
1           -0.060590               -0.70889                 3.9905   
2           -0.037177                1.45710                 2.2828   
3           -0.035904                0.71096                 1.8582   
4           -0.015314                3.39960                 2.7881   

   linear_acceleration_Z  
0                -5.5439  
1                -8.0273  
2               -11.2990  
3               -12.2270  
4               -10.4100

Here we see the different measurements used for the sensor data. Our goal is to predict what type of surface the robot is driving on. We expect that the different surfaces will make the robot shake and bump around in subtle ways so we can look at frequency graphs for each measurement channel. We will also look at the groupings and classes a bit more to see what is going on as it is strange that data would have groups.

This dataset was for a competition and the group labels can be used to find the correct labels for the test data by expanding the groups and using time series forcasting to see which group the test data belongs too. This does not involve machine learning or data science at all though so we will not be doing that. If you want to see a method that does do it you can look at this post here https://www.kaggle.com/trohwer64/submission-fourier-neighbour-detection-svm

In [160]:
# First lets explore the data some.
ytrain.groupby(['surface']).agg(['count'])

series_id group_id
                           count    count
surface                                  
carpet                       189      189
concrete                     779      779
fine_concrete                363      363
hard_tiles                    21       21
hard_tiles_large_space       308      308
soft_pvc                     732      732
soft_tiles                   297      297
tiled                        514      514
wood                         607      607

In [161]:
# We see that there are 9 classes and the data is VERY imbalanced. Lets look at how the groups are set up
y = ytrain.groupby(['group_id', 'surface']).agg(['count'])
with pd.option_context("display.max_rows", 1000):
    display(y.sort_values(['surface', 'group_id']))

series_id
                                    count
group_id surface                         
7        carpet                        56
25       carpet                        51
52       carpet                        11
60       carpet                        71
0        concrete                      57
5        concrete                       9
12       concrete                      57
20       concrete                      57
31       concrete                      54
32       concrete                      57
39       concrete                      70
41       concrete                      70
42       concrete                      70
47       concrete                      70
50       concrete                      10
57       concrete                      41
61       concrete                      70
62       concrete                      70
63       concrete                      17
13       fine_concrete                 57
30       fine_concrete                  3
35       fine_concrete                 57
36       fine_concrete                 36
48       fine_concrete                 70
68       fine_concrete                 70
72       fine_concrete                 70
27       hard_tiles                    21
16       hard_tiles_large_space        45
33       hard_tiles_large_space        57
40       hard_tiles_large_space        71
44       hard_tiles_large_space        71
59       hard_tiles_large_space        64
3        soft_pvc                      57
6        soft_pvc                      48
18       soft_pvc                      57
19       soft_pvc                      57
26       soft_pvc                      57
29       soft_pvc                      57
34       soft_pvc                      57
37       soft_pvc                      34
51       soft_pvc                      60
53       soft_pvc                      71
56       soft_pvc                      29
58       soft_pvc                       7
69       soft_pvc                      70
70       soft_pvc                      71
4        soft_tiles                    57
9        soft_tiles                    57
15       soft_tiles                    57
17       soft_tiles                    12
21       soft_tiles                    57
22       soft_tiles                    57
1        tiled                         38
8        tiled                         57
11       tiled                         56
28       tiled                         36
54       tiled                         70
55       tiled                         70
64       tiled                         53
66       tiled                         64
71       tiled                         70
2        wood                          18
10       wood                          56
14       wood                          56
23       wood                          56
24       wood                           5
38       wood                          70
43       wood                          70
45       wood                          70
46       wood                          70
49       wood                          60
65       wood                          70
67       wood                           6

In [ ]:
# Wow! The data is all over the place! There are a total of 73 groups with wildly different amounts of series in each group.
# Next lets look at some graphs of the data.

